File: golang/alexa/handlers.go
package alexa

import (
	"alexa-llm-go/llm"
	"embed"
	"encoding/json"
	"fmt"
	"log"
)

//go:embed templates/*.json
var templatesFS embed.FS

type AlexaHandler struct {
	LLMClient *llm.OpenRouterClient
}

func NewAlexaHandler(llmClient *llm.OpenRouterClient) *AlexaHandler {
	return &AlexaHandler{LLMClient: llmClient}
}

func (h *AlexaHandler) Handle(req RequestEnvelope) ResponseEnvelope {
	resp := ResponseEnvelope{
		Version:           "1.0",
		SessionAttributes: req.Session.Attributes,
	}
	if resp.SessionAttributes == nil {
		resp.SessionAttributes = make(map[string]interface{})
	}

	switch req.Request.Type {
	case "LaunchRequest":
		h.handleLaunchRequest(&req, &resp)
	case "IntentRequest":
		switch req.Request.Intent.Name {
		case "AskIntent":
			h.handleAskIntent(&req, &resp)
		case "AMAZON.CancelIntent", "AMAZON.StopIntent":
			h.handleCancelOrStopIntent(&req, &resp)
		default:
			h.handleFallback(&req, &resp)
		}
	case "SessionEndedRequest":
		h.handleSessionEnded(&req, &resp)
	default:
		h.handleFallback(&req, &resp)
	}

	return resp
}

func (h *AlexaHandler) setSpeak(resp *ResponseEnvelope, text string, endSession bool) {
	resp.Response.OutputSpeech = &OutputSpeech{
		Type: "PlainText",
		Text: text,
	}
	resp.Response.ShouldEndSession = endSession
}

func (h *AlexaHandler) setAsk(resp *ResponseEnvelope, txt string, repromptTxt string) {
	h.setSpeak(resp, txt, false)
	resp.Response.Reprompt = &Reprompt{
		OutputSpeech: OutputSpeech{
			Type: "PlainText",
			Text: repromptTxt,
		},
	}
}

func (h *AlexaHandler) addAPLDirective(resp *ResponseEnvelope, templateName string, datasources map[string]interface{}) {
	templateData, err := templatesFS.ReadFile("templates/" + templateName)
	if err != nil {
		log.Printf("Error reading template file %s: %v", templateName, err)
		return
	}

	var document map[string]interface{}
	if err := json.Unmarshal(templateData, &document); err != nil {
		log.Printf("Error decoding template %s: %v", templateName, err)
		return
	}

	directive := DirectiveAPL{
		Type:        "Alexa.Presentation.APL.RenderDocument",
		Document:    document,
		Datasources: datasources,
	}
	resp.Response.Directives = append(resp.Response.Directives, directive)
}

func (h *AlexaHandler) handleLaunchRequest(req *RequestEnvelope, resp *ResponseEnvelope) {
	speakOutput := "Olá! Sou seu assistente inteligente. O que deseja me perguntar hoje?"
	resp.SessionAttributes["chat_history"] = []llm.Message{}

	if req.Context.System.Device.SupportsAPL() {
		h.addAPLDirective(resp, "welcome_template.json", map[string]interface{}{
			"welcomeData": map[string]interface{}{
				"header": "Bem-vindo!",
				"text":   "Estou pronto para ouvir. Faça sua pergunta.",
			},
		})
	}

	h.setAsk(resp, speakOutput, speakOutput)
}

func (h *AlexaHandler) handleAskIntent(req *RequestEnvelope, resp *ResponseEnvelope) {
	userQuery := "O que você pode fazer?"
	if slot, ok := req.Request.Intent.Slots["query"]; ok && slot.Value != "" {
		userQuery = slot.Value
	}

	history := GetChatHistory(req.Session.Attributes)
	var answer string
	if h.LLMClient != nil {
		answer, history = h.LLMClient.Ask(userQuery, history)
	} else {
		answer = "O cliente LLM não está configurado."
		history = append(history, llm.Message{Role: "user", Content: userQuery}, llm.Message{Role: "assistant", Content: answer})
	}

	resp.SessionAttributes["chat_history"] = history

	if req.Context.System.Device.SupportsAPL() {
		h.addAPLDirective(resp, "response_template.json", map[string]interface{}{
			"chatData": map[string]interface{}{
				"question": userQuery,
				"answer":   answer,
			},
		})
	}

	h.setAsk(resp, answer, "Deseja me perguntar mais alguma coisa?")
}

func (h *AlexaHandler) handleCancelOrStopIntent(req *RequestEnvelope, resp *ResponseEnvelope) {
	h.setSpeak(resp, "Até logo! Estarei aqui se precisar.", true)
}

func (h *AlexaHandler) handleSessionEnded(req *RequestEnvelope, resp *ResponseEnvelope) {
	resp.Response.ShouldEndSession = true
}

func (h *AlexaHandler) handleFallback(req *RequestEnvelope, resp *ResponseEnvelope) {
	intentName := req.Request.Intent.Name
	var speakOutput string
	if intentName != "" {
		speakOutput = fmt.Sprintf("A intenção %s foi chamada, mas não tenho um handler para ela.", intentName)
	} else {
		speakOutput = fmt.Sprintf("Recebi um evento do tipo %s, que não sei como processar.", req.Request.Type)
	}
	h.setAsk(resp, speakOutput, speakOutput)
}

File: python/src/alexa/handlers.py
import json
import os
from ask_sdk_core.dispatch_components import AbstractRequestHandler, AbstractRequestInterceptor
from ask_sdk_core.utils import is_request_type, is_intent_name
from ask_sdk_core.handler_input import HandlerInput
from ask_sdk_model import Response
from ask_sdk_core.exceptions import AlexaSdkException
from ask_sdk_model.interfaces.alexa.presentation.apl import RenderDocumentDirective
from src.llm.client import llm_client

class LaunchRequestHandler(AbstractRequestHandler):
    """Handler para quando a skill é aberta ('Alexa, abra meu assistente')."""
    def can_handle(self, handler_input: HandlerInput) -> bool:
        return is_request_type("LaunchRequest")(handler_input)

    def handle(self, handler_input: HandlerInput) -> Response:
        speak_output = "Olá! Sou seu assistente inteligente. O que deseja me perguntar hoje?"
        
        # Iniciando histórico na sessão (Memória que vovê pediu)
        session_attr = handler_input.attributes_manager.session_attributes
        session_attr["chat_history"] = []
        handler_input.attributes_manager.session_attributes = session_attr

        # Adicionando visual APL se for um Echo Show
        if handler_input.request_envelope.context.system.device.supported_interfaces.alexa_presentation_apl:
            apl_document = self._load_apl_template("welcome_template.json")
            handler_input.response_builder.add_directive(
                RenderDocumentDirective(
                    document=apl_document,
                    datasources={
                        "welcomeData": {
                            "header": "Bem-vindo!",
                            "text": "Estou pronto para ouvir. Faça sua pergunta."
                        }
                    }
                )
            )

        return (
            handler_input.response_builder
                .speak(speak_output)
                .ask(speak_output) # Mantém a skill aberta esperando resposta
                .response
        )

    def _load_apl_template(self, filename: str):
        path = os.path.join(os.path.dirname(__file__), "templates", filename)
        with open(path, "r") as f:
            return json.load(f)

class AskIntentHandler(AbstractRequestHandler):
    """Handler principal que recebe a pergunta do usuário e envia para a LLM."""
    def can_handle(self, handler_input: HandlerInput) -> bool:
        return is_intent_name("AskIntent")(handler_input)

    def handle(self, handler_input: HandlerInput) -> Response:
        # Pegar a pergunta do slot 'query' (precisa configurar no console da Alexa)
        slots = handler_input.request_envelope.request.intent.slots
        user_query = slots["query"].value if "query" in slots else "O que você pode fazer?"

        # Recuperar histórico da sessão (Contexto)
        session_attr = handler_input.attributes_manager.session_attributes
        history = session_attr.get("chat_history", [])

        # Chamar a LLM
        answer = llm_client.ask(user_query, history=history)

        # Atualizar histórico com a nova interação (Contexto de Memória)
        history.append({"role": "user", "content": user_query})
        history.append({"role": "assistant", "content": answer})
        session_attr["chat_history"] = history
        handler_input.attributes_manager.session_attributes = session_attr

        # Preparar visual APL
        if handler_input.request_envelope.context.system.device.supported_interfaces.alexa_presentation_apl:
            apl_document = self._load_apl_template("response_template.json")
            handler_input.response_builder.add_directive(
                RenderDocumentDirective(
                    document=apl_document,
                    datasources={
                        "chatData": {
                            "question": user_query,
                            "answer": answer
                        }
                    }
                )
            )

        return (
            handler_input.response_builder
                .speak(answer)
                .ask("Deseja me perguntar mais alguma coisa?") 
                .response
        )

    def _load_apl_template(self, filename: str):
        path = os.path.join(os.path.dirname(__file__), "templates", filename)
        with open(path, "r") as f:
            return json.load(f)

class CancelOrStopIntentHandler(AbstractRequestHandler):
    def can_handle(self, handler_input: HandlerInput) -> bool:
        return (is_intent_name("AMAZON.CancelIntent")(handler_input) or
                is_intent_name("AMAZON.StopIntent")(handler_input))

    def handle(self, handler_input: HandlerInput) -> Response:
        return (
            handler_input.response_builder
                .speak("Até logo! Estarei aqui se precisar.")
                .set_should_end_session(True)
                .response
        )

class SessionEndedRequestHandler(AbstractRequestHandler):
    def can_handle(self, handler_input: HandlerInput) -> bool:
        return is_request_type("SessionEndedRequest")(handler_input)

    def handle(self, handler_input: HandlerInput) -> Response:
        # Apenas aceita sem erro
        return handler_input.response_builder.response

class CatchAllRequestHandler(AbstractRequestHandler):
    """Fallback genérico para intenções ou eventos que não configuramos."""
    def can_handle(self, handler_input: HandlerInput) -> bool:
        return True

    def handle(self, handler_input: HandlerInput) -> Response:
        req = handler_input.request_envelope.request
        if hasattr(req, "intent"):
            intent_name = req.intent.name
            speak_output = f"A intenção {intent_name} foi chamada, mas não tenho um handler para ela."
        else:
            speak_output = f"Recebi um evento do tipo {req.type}, que não sei como processar."
            
        return (
            handler_input.response_builder
                .speak(speak_output)
                .ask(speak_output)
                .response
        )

class SkillIdVerifierInterceptor(AbstractRequestInterceptor):
    """Verifica se o ID da Skill que está chamando está na lista permitida."""
    def process(self, handler_input):
        # type: (HandlerInput) -> None
        expected_ids_str = os.getenv("ALEXA_SKILL_ID")
        if not expected_ids_str:
            return

        allowed_ids = [s.strip() for s in expected_ids_str.split(",")]
        current_id = handler_input.request_envelope.context.system.application.application_id
        
        if current_id not in allowed_ids:
            print(f"Acesso negado: Skill ID {current_id} não autorizado.")
            raise AlexaSdkException("Skill ID não autorizado.")
